api/proto/teleport/legacy/types/types.proto

```diff
@@ -1892,14 +1897,155 @@ message ProvisionTokenSpecV2Oracle {
   repeated Rule Allow = 1 [(gogoproto.jsontag) = "allow,omitempty"];
 }
 
+// ProvisionTokenSpecV2BoundKeypair contains configuration for bound-keypair
+// type join tokens.
+message ProvisionTokenSpecV2BoundKeypair {
+  // OnboardingSpec contains parameters for initial joining and keypair
+  // registration.
+  message OnboardingSpec {
+    // InitialPublicKey is an initial, static public key that will be used to
+    // immediately populate `.status.bound_keypair.bound_public_key`. If set,
+    // `initial_join_secret` and `must_join_before` are ignored, and
+    // `.status.bound_keypair.initial_join_secret` will not be generated. This
+    // value is written in SSH public key format, including algorithm.
+    string InitialPublicKey = 1 [(gogoproto.jsontag) = "initial_public_key,omitempty"];
+
+    // InitialJoinSecret is an initial join secret joining clients may use to
+    // register their public key on first join. If `initial_public_key` is set,
+    // this value is ignored. Otherwise, if set, this value will be used to
+    // populate `.status.bound_keypair.intitial_join_secret`. If unset and no
+    // `initial_public_key` is provided, a random secure value will be generated
+    // server-side to populate the status field.
+    string InitialJoinSecret = 2 [(gogoproto.jsontag) = "initial_join_secret,omitempty"];
+
+    // MustJoinBefore is an optional time before which initial secret joining
+    // may be used. Attempts to register using an initial join secret after this
+    // timestamp will not be allowed. This may be modified after creation if
+    // necessary to allow the initial registration to take place.
+    google.protobuf.Timestamp MustJoinBefore = 3 [
+      (gogoproto.stdtime) = true,
+      (gogoproto.nullable) = true,
+      (gogoproto.jsontag) = "must_join_before,omitempty"
+    ];
+  }
```

Comment:
From today's discussion: rename this to more explicitly associate it with `initial_join_secret`; possibly group them into their own sub-message. I'll play with the ergonomics of that today.

Reply:
This field has been renamed to `must_register_before`